Suggest product categories with AI

nima-karimi opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR adds the functionality to suggest product categories using AI. Currently, it can generate suggestions based on the available categories only. Meaning that it will check the list of available categories in the store and suggest which ones are a good match for the product.

Once suggestions are returned, a list is displayed to the user where they can select to add each suggestion.

The reason why it's only suggesting existing categories instead of allowing the user to create new categories is because we do not have an easy method to create multi-level categories (Parent > Child) on the product edit page through JS. This can be revisited in another PR, but since the amount of work to implement it was substantial, I skipped it for the MVP phase.

As part of #39436

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new product or edit an existing one
2. Make sure you have some categories defined in your store
3. Enter some data (title, description, tags, etc.) for your product
4. Click "Suggest categories using AI" on the category meta box
5. Confirm that the categories returned are a good match for the product based on your existing categories
6. Select some categories to add them to your product
7. Update the product and confirm the categories will be added to it
8. Edit the product again and click on the suggest categories button
9. This time, it should not return any categories that you previously selected (or are already assigned to the product). It should instead display a message indicating that your product is well-categorized.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

nima-karimi committed:
Accept arguments for the TinyContent getContent method.

Used to fetch the plain text version of the description.

nima-karimi committed:
Add a class to the loading message content

nima-karimi committed:
Set the max description length as global constant

nima-karimi committed:
Create a MagicButton component and use it

nima-karimi committed:
Get plain text description instead of HTML

nima-karimi committed:
Return full category hierarchy (Parent > Child)

nima-karimi committed:
Add method to return all available categories on the product edit page

nima-karimi committed:
Helper function to generate product data instructions for the prompt

nima-karimi committed:
Helper function to select category checkboxes on DOM

nima-karimi committed:
Create component to display a list of suggestion items as pills

nima-karimi committed:
Add product category suggestions to product edit page

nima-karimi committed:
Merge branch 'trunk' into feature/suggest-product-categories-39436

# Conflicts:
#	plugins/woo-ai/src/product-description/product-description-button-container.tsx

nima-karimi committed:
Use the AI package to get text completion

nima-karimi committed:
Add tracks

nima-karimi committed:
Add changelog

nima-karimi commented:
@jarekmorawski Could you please check the UI/UX for this feature and let me know if you have any suggestions? Thanks.

jarekmorawski commented:
Absolutely. I don't have a working dev environment. Would you mind building a zip or setting up a [playground](https://developer.wordpress.org/playground/)?

nima-karimi commented:
Thanks, @jarekmorawski. I've attached a zip file for the build:
[woo-ai.zip](https://github.com/woocommerce/woocommerce/files/12177341/woo-ai.zip)

tommyshellberg commented:
Just wanted to paste in the comments I had from Slack so they don't get lost in the backscroll:

### How to update the categories list.

Nima said:
> I could try calling the API directly and creating all these but then I’m not sure how to update the categories list that’s already on the page.

> The categories are loaded via a PHP call and then statically loaded on the page. However, the product categories UI in the product editor does have logic to refresh the list after creating a new one - it looks like that’s through the add-product_cat AJAX action.
It looks like it takes care of reloading the UI too:

![Screenshot 2023-07-27 at 16 15 59](https://github.com/woocommerce/woocommerce/assets/6723003/29a2f4fd-6b3c-4e73-803c-50fb94cb63e6)

### Using the WC REST API

>I think this could be done using the Woo Categories batch update endpoint - you could create multiple categories first with a single API call (don’t associate them with each other), and in the return of the API call you’ll get their IDs.
You make a second API call, using those IDs to create the hierarchical relationship (update the child ID to set its parent to the parent ID).
https://woocommerce.github.io/woocommerce-rest-api-docs/#batch-update-product-categories

nima-karimi commented:
Thanks, @tommyshellberg. I tried using the AJAX endpoint, but based on what I gathered, its output is in XML and is fed to a JS element called WP List, and it all happens automatically. 

We can call the endpoint manually, but then the list wouldn't be updated, so we'd have to parse the XML response and update the list on the DOM manually. It might work, but it just feels like a dirty workaround IMO.

We can insert the new category by manually filling out the new category form and submitting it. This would trigger a call to the endpoint and update the list automatically based on the response. Still, we wouldn't know the ID of the newly added category (since we won't get the response back), so we can't select it from the list.

The WP List element has an `addAfter` setting that accepts a callback; we could use it to get the ID of the newly added category, but the callback for that setting can only be set when initializing the list. I couldn't find a way to change it by accessing the wpList object.

We could also add the new category with the REST API and then use the wpList's `add` method to update the list but I didn't get deeper into how the `add` function works for wpList. I suspect it accepts an XML input, which means we'd still have to map the new category into an XML before updating the list.

I think we can explore this in another issue too, and ship this part first and see how the usage is. Let me know what you think.

nima-karimi committed:
Merge branch 'trunk' into feature/suggest-product-categories-39436

# Conflicts:
#	plugins/woo-ai/src/product-description/product-description-button-container.tsx
#	plugins/woo-ai/src/utils/tiny-tools.ts

nima-karimi committed:
Fix merge conflict

tommyshellberg commented on plugins/woo-ai/src/components/suggestion-pills/suggestion-pill-item.tsx:
```diff
@@ -0,0 +1,25 @@
+	onSuggestionClick: ( suggestion: string ) => void;
+};
+
+export const SuggestionPillItem: React.FC< SuggestionPillItemProps > = ( {
```
These look nice, great work here!

tommyshellberg commented on plugins/woo-ai/src/index.js:
```diff
@@ -32,6 +33,16 @@ const renderComponent = ( Component, rootElement ) => {
+	renderComponent( ProductCategorySuggestions, root );
+
+	// Insert the category suggestions node in the product category meta box.
+	document.getElementById( 'taxonomy-product_cat' ).append( root );
```
Was it not possible to add the app root div element using a WordPress PHP filter?

nima-karimi replied:
Unfortunately no. I couldn't find any filters that applied to the categories' meta box (or any taxonomy meta box).

tommyshellberg replied:
So I said previously that you could use `renderComponent()`... and you are, indeed!
This means the suggestions component automatically has a react query client provider and can use react query to decide when to make api calls like here:

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woo-ai/src/hooks/useStoreBranding.ts#L33

Could be an enhancement issue.

tommyshellberg commented on plugins/woo-ai/src/product-category/product-category-suggestions.tsx:
```diff
@@ -0,0 +1,209 @@
+			"Your goal is to enhance the store's SEO performance and sales.",
+			'Categories can have parents and multi-level children structures like Parent Category > Child Category.',
+			'You will be given a list of available categories and you can only return the best matching categories from this list.',
+			`Available categories are: ${ allAvailableCategories }`,
```
Have you experimented with not giving the prompt the available categories? For instance, ChatGPT might be able to suggest categories based on the [Google product taxonomy hierarchy](https://www.google.com/basepages/producttype/taxonomy.en-US.txt), which I think is a good idea to use as the basis for categories.
I tested essentially this same exact prompt but changed it so that it would try to match based on the google product taxonomy hierarchy:


> You are a WooCommerce SEO and marketing expert.
> Using the product's name, description, categories, detect the main product category and its subcategories based on the existing available categories.
> Your goal is to enhance the store's SEO performance and sales.
> Categories can have parents and multi-level children structures like Parent Category > Child Category.
> Use the Google standard taxonomy hierarchy to suggest categories.
> The product's properties are:
> Name: High-Quality Espresso Machine for Home Use.
> Description: The High-Quality Espresso Machine for Home Use by Shellberg Books is the perfect companion for all those caffeine lovers out there! This incredible machine is designed to bring the café experience right to your kitchen, so you can have the perfect cup of espresso whenever you desire.
> Made with utmo.
> Product categories: Appliances.
> Return only a comma-separated list of the product categories, children categories must be separated by >.
> Here is an example of a valid response:
> Parent Category 1 > Subcategory 1, Parent Category 2 > Subcategory 2 > Another Subcategory 2
> Do not output the example response. Respond only with the suggested categories. Do not say anything else.

Here was the response (GPT3.5):
![Screenshot 2023-08-11 at 15 28 12](https://github.com/woocommerce/woocommerce/assets/6723003/c357b544-3cbf-459d-a2d4-7c5c974d6722)

What I like about this idea:

- The prompt can be a lot shorter, especially if the store has a lot of categories.
- We can actually suggest categories that don't exist, even if we can't create them automatically. We could say something like:
> The recommended category for this product is "Appliances > Kitchen Appliances > Coffee & Tea Appliances" but it doesn't appear to exist yet. Please create the category so we can suggest it for future products."

What do you think?

tommyshellberg replied:
I had to ask, just to be sure... :) 
![Screenshot 2023-08-11 at 17 04 07](https://github.com/woocommerce/woocommerce/assets/6723003/479872fa-e130-4183-8190-37d8a9ed9349)

nima-karimi replied:
I like this idea. My main reason for going with the available categories was that we reduce friction by selecting what they have available, and it would be a good enough MVP. We can certainly add this feature in the next version, though. If we can create the categories automatically, it would be much better. 

I'm unsure how it would look if we just suggested categories and asked them to create them. If I were the user, my first reaction would be why the smart, artificially intelligent app can process all the data in my store but can't just create the categories for me 😄

tommyshellberg commented:
This seems to work pretty well when the categories already exist, nice work! 👏 
My test described here is for an Espresso Machine product.
I created two categories, Appliances->Coffee Machines and it was able to suggest both the child and parent categories:
![Screenshot 2023-08-11 at 15 02 02](https://github.com/woocommerce/woocommerce/assets/6723003/afb6ffd0-187d-4479-8f6a-0abc2afba5be)

Before I created those categories, it was unable to match it to any category:
![Screenshot 2023-08-11 at 14 47 26](https://github.com/woocommerce/woocommerce/assets/6723003/daefc412-797a-4ef0-97d0-f4dcfe0cdc9e)
What do you think about changing the message in this case? The product's category assigned was "uncategorized" when I tried to generate category suggestions so I think a message more like "We were unable to recommend a category for this product. Please create the category so we can suggest it for future products".

I also left a comment about possibly not injecting the available categories in the prompt and instead having GPT try to match based on the https://github.com/woocommerce/woocommerce/pull/39437/files#r1291343200.

tommyshellberg commented on plugins/woo-ai/src/product-category/product-category-suggestions.tsx:
```diff
@@ -0,0 +1,209 @@
+					</p>
+				</div>
+			) }
+			{ suggestionsState === SuggestionsState.None && (
```
I wonder if we could simplify this by combining "NoMatch and None" into a single "None" state. Ultimately, it seems like the same result: the user doesn't get a category suggestion, but for slightly different reasons. IMO we could combine them into a single state and make the returned message more generic, maybe like "Unable to generate a matching category for the product." 🤔

nima-karimi replied:
Yeah that makes sense. I've removed the nomatch state in https://github.com/woocommerce/woocommerce/pull/39437/commits/d1c28b12f0662fe38e44ff464004d8c615ddcf41

tommyshellberg commented on plugins/woo-ai/src/utils/productData.ts:
```diff
@@ -125,3 +178,47 @@ export const isProductVirtual = () =>
+ *
+ * @return {string[]} Array of category names
+ */
+export const getAllAvailableCategories = (): string[] => {
```
What if we re-work this to make a REST API call to the Woo categories endpoint? It seems like it would be a lot less brittle, and, it could be used outside of the classic product editor (the beta blocks-based editor).
https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-product-categories

nima-karimi replied:
I'd say let's also push this into the backlog for when we need this functionality in other areas. Each approach has pros and cons, and I think even if we use the API, we still have to go through each item in the list and map the parent > child relationships, and we'd then have a delay and potential failure for the API request as well.

nima-karimi committed:
Remove NoMatch state for category suggestions

nima-karimi committed:
Merge branch 'trunk' into feature/suggest-product-categories-39436

# Conflicts:
#	plugins/woo-ai/src/index.scss

nima-karimi commented:
Thanks for the review, @tommyshellberg. I tried to address your comments and made some changes to the code. Ready for another review 🙏

nima-karimi commented:
Todos (based on our meeting today): 

- [ ] Change the error message when the AI cannot find category matches to state the returned categories clearly and indicate why it cannot (because they are already selected or because there's not enough product information, etc.).
- [ ] Get the list of available categories from the PHP side and inject it into react (via a JS global variable) instead of going through the DOM and fetching the types.
- [ ] Return new categories that might be relevant to the product and list them on the page as suggestions so that the user can add them manually if they want (this is a workaround for now for the issue where we can't add new categories automatically).

tommyshellberg commented:
Are we sure that we can't grab the categories using a WC REST API request?
<img width="724" alt="Screenshot 2023-08-17 at 22 19 09" src="https://github.com/woocommerce/woocommerce/assets/6723003/35898abc-d356-4659-b020-8c94527d68b9">

nima-karimi commented:
> Are we sure that we can't grab the categories using a WC REST API request?

We certainly can. I guess the only difference would be the delay to get the response, which _can_ be acceptable since the LLM call takes much longer. We also need to add some error handling for when the API request fails. I tried to avoid getting into that for this release and push it back to the next version but since I'm already refactoring that part, might as well do it properly.

I can also add another API to the backend that returns the categories with their parents included so that we won't have to process this on the browser. All those recursive calls (if the store has many complex categories) feel more like a backend rather than a browser task.

nima-karimi committed:
Merge branch 'trunk' into feature/suggest-product-categories-39436

nima-karimi committed:
Get available categories using WC REST API

nima-karimi committed:
Suggest new categories

nima-karimi committed:
Run separate prompts for existing and new category generation

nima-karimi committed:
Fix overflow in suggestion pills

nima-karimi committed:
Don't include existing selected categories in prompt

nima-karimi committed:
Add util to encode html entities

nima-karimi committed:
Exclude "Uncategorized" category from product data

nima-karimi committed:
Allow excluding properties from the product data instructions

nima-karimi committed:
Create category from suggestion if it doesn't exist

nima-karimi committed:
Merge branch 'trunk' into feature/suggest-product-categories-39436

tommyshellberg commented:
> I've modified this PR significantly, so I think it's best if you do another review from the top (sorry 😅).

No problem, I do have to say that I've just merged a couple of PRs so there are now conflicts to address (sorry!), I'll review the code while you fix the conflicts in parallel. 👍

nima-karimi committed:
Show suggestions as links instead of pills

nima-karimi committed:
Ask for feedback after suggestion selected

nima-karimi committed:
Merge branch 'trunk' into feature/suggest-product-categories-39436

# Conflicts:
#	plugins/woo-ai/src/components/description-completion-buttons.tsx
#	plugins/woo-ai/src/components/index.ts
#	plugins/woo-ai/src/constants.ts
#	plugins/woo-ai/src/index.js
#	plugins/woo-ai/src/index.scss
#	plugins/woo-ai/src/product-description/product-description-button-container.tsx
#	plugins/woo-ai/src/utils/tiny-tools.ts

nima-karimi committed:
Decode html entities in available categories results

nima-karimi commented:
> The Pill component which contains each recommendation can get pretty awkward once the taxonomy gets into multiple levels:

Yeah, I couldn't figure out a clean way to resolve that issue. I changed it to look like a link. But I think it's just too much text on there now, which can be a bit confusing. I think we need to do a redesign on this later when @jarekmorawski comes back. Do you think it looks good enough for an MVP, though?

> the category it returned doesn’t seem to truly match up with a Google Taxonomy.

Yup! I tried GPT-4 as well, but even though the results were more consistent, I still got the same categories that simply don't exist in the list of Google categories. I think the LLM is just way too creative to stick to that list. We probably need to train a specific model or use a vector DB to narrow down the list of categories that it can choose from and send it in the prompt.

> render the CategoriesSuggestion component using renderComponent

I like this idea, but let's pencil it down for later because this PR has been open long enough 😅 We can add it as an improvement late. 

>  I did see some characters that were still displayed as their encoded versions:

Yeah, that's a side effect of using the categories API to fetch the list of available categories. It encodes the HTML entities in the JSON response, and we pass them as is to the LLM. The LLM returns the encoded characters, too. I added some code to decode those characters. I was a little hesitant to add decoding because of XSS issues, but I think the solution with textarea should resolve the XSS issue.

nima-karimi commented:
@tommyshellberg Ready for another review 🙏 

I've also added a feedback box that appears under the category suggestions whenever the user clicks a suggestion. I didn't want the box to appear every time so I added a dismiss button there as well as a "dismiss forever" button so that the user can hide it if it gets annoying. It takes a huge space under the categories box so it's not looking very pretty right now. Let me know what you think.

tommyshellberg commented:
> I like this idea, but let's pencil it down for later because this PR has been open long enough

What if we merge this PR into a feature branch rather than trunk? We could then add the remaining improvements as issues and merge into the feature branch. I know that working on a large PR can cause a lot of fatigue and it would make splitting up the improvements easier. 

One of the things I'm thinking about, for instance, is if we should even use an LLM for this. Given that they tend to hallucinate, and given that these taxonomies rarely change (the last update was in 2019?), we could just use a fuzzy search algorithm with the taxonomies hard-coded. There's a library we could take inspiration from here:
https://github.com/D3VL/google-taxonomy-matcher/tree/master

tommyshellberg commented:
> . I think we need to do a redesign on this later when @jarekmorawski comes back. Do you think it looks good enough for an MVP, though?

I think it would be worth experimenting with truncating all of the parent categories and only showing within the pill the lowest (the most relevant) category:
![Screenshot 2023-09-07 at 15 55 57](https://github.com/woocommerce/woocommerce/assets/6723003/5c55a44c-63e6-4c9b-8ae6-ee3304b1a701)

You could do something like show the whole hierarchy when hovering over the pill button.
But, as mentioned, I'm okay with merging this into a feature branch so we can split up the remaining issues into smaller PRs.

tommyshellberg commented on plugins/woo-ai/src/product-category/category-suggestion-feedback.tsx:
```diff
@@ -0,0 +1,129 @@
+	};
+
+	// hide the feedback notice and store the user's preference in local storage so we don't show it again
+	const hideFeedbackBoxForever = () => {
```
I created a persistent preferences store for the plugin which is initialized in index.js through setPreferencesPersistence(). You could use that for dismissing notices as I used for the notice here:
https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woo-ai/src/components/info-modal/info-modal.tsx#L36

tommyshellberg replied:
Feel free to add that as an enhancement if we end up merging to a feature branch.

tommyshellberg commented on plugins/woo-ai/src/product-category/category-suggestion-feedback.tsx:
```diff
@@ -0,0 +1,129 @@
+ */
+import { recordCategoryTracks } from './utils';
+
+const WOO_AI_HIDE_CATEGORY_SUGGESTION_FEEDBACK =
```
I would actually be okay with cutting the feedback collector out of scope for this PR so we can get some ideas from design; we may end up re-using the feedback snackbar that we've been using, or go with something closer to the categories meta box.

nima-karimi committed:
Don't encode html entities when comparing available categories

nima-karimi committed:
Change feedback box style

nima-karimi committed:
Suggest only one category

nima-karimi committed:
Remove log

nima-karimi committed:
Show feedback box after generating suggestions

Instead of showing it after a suggestion is selected

nima-karimi commented:
@tommyshellberg This is ready for another review 🙏 I've made some changes according to our dev call.

nima-karimi committed:
Fix typo

tommyshellberg commented on plugins/woo-ai/src/product-category/product-category-suggestions.tsx:
```diff
@@ -0,0 +1,372 @@
+					! existingSuggestions.includes( suggestion )
+			);
+		} catch ( e ) {
+			// eslint-disable-next-line no-console
```
Do we want a Tracks event for this case, where we can't fetch categories? It would be nice to have visibility into this.

tommyshellberg reviewed (approved):
Thanks for working on this and grinding through it, let's get it into the hands of folks to try it out. 👍

tommyshellberg commented:
This is looking much more compact, I like it. Nice work! 

Having the feedback as a smaller component makes it look much better:
![Screenshot 2023-09-12 at 14 37 51](https://github.com/woocommerce/woocommerce/assets/6723003/c92201f3-3ef3-48fa-99d2-22e3fe413fc4)

One of your commit messages is named `Suggest only one category`, is this intended to only ever show one recommendation, or only one _new_ recommendation? At the moment, if I have existing categories, it will show multiple suggestions.
I think this is fine for now, just wanted to clarify.

nima-karimi commented:
> is this intended to only ever show one recommendation, or only one new recommendation?

Oh yeah, that's a bit misleading. Sorry about that. So, the intention was to show one recommendation per type. I wish there was an easy way to choose the best one out of the two recommendations (new categories and existing ones) and simply show that to the user. We can send them both back to the LLM and ask for the best one or just run an algorithm on our end to see which one is more relevant to the text. Perhaps we can add this as an improvement later.
